libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/WriteOperation.java

```diff
@@ -26,9 +26,27 @@
  * operation is that is has a listener and a reference to its channel. The actual conversion of the write
  * operation implementation to bytes will be performed by the {@link SocketChannelContext}.
  */
-public interface WriteOperation {
+public class WriteOperation {
 
-    BiConsumer<Void, Throwable> getListener();
+    private final SocketChannelContext channelContext;
+    private final Object writeObject;
+    private final BiConsumer<Void, Throwable> listener;
 
-    SocketChannelContext getChannel();
+    public WriteOperation(SocketChannelContext channelContext, Object writeObject, BiConsumer<Void, Throwable> listener) {
```

Comment:
why do we have to loose all type safety here. Can't we fix the interface to return `FlushOperation` and return this?